agent-codeless/src/main/java/com/microsoft/applicationinsights/agentc/internal/diagnostics/status/StatusFile.java

```diff
@@ -0,0 +1,197 @@
+package com.microsoft.applicationinsights.agentc.internal.diagnostics.status;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.microsoft.applicationinsights.agentc.internal.diagnostics.AgentExtensionVersionFinder;
+import com.microsoft.applicationinsights.agentc.internal.diagnostics.DiagnosticsHelper;
+import com.microsoft.applicationinsights.agentc.internal.diagnostics.DiagnosticsValueFinder;
+import com.microsoft.applicationinsights.agentc.internal.diagnostics.InstrumentationKeyFinder;
+import com.microsoft.applicationinsights.agentc.internal.diagnostics.MachineNameFinder;
+import com.microsoft.applicationinsights.agentc.internal.diagnostics.PidFinder;
+import com.microsoft.applicationinsights.agentc.internal.diagnostics.SdkVersionFinder;
+import com.microsoft.applicationinsights.agentc.internal.diagnostics.SiteNameFinder;
+import com.microsoft.applicationinsights.internal.system.SystemInformation;
+import com.microsoft.applicationinsights.internal.util.ThreadPoolUtils;
+import com.squareup.moshi.Moshi.Builder;
+import okio.BufferedSink;
+import okio.Okio;
+import okio.Sink;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.io.File;
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.RuntimeMXBean;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class StatusFile {
+
+    private StatusFile() {}
+
+    private static final List<DiagnosticsValueFinder> VALUE_FINDERS = new ArrayList<>();
+
+    @VisibleForTesting
+    static final Map<String, Object> CONSTANT_VALUES = new ConcurrentHashMap<>();
+
+    @VisibleForTesting
+    static final String FILENAME_PREFIX = "status";
+
+    @VisibleForTesting
+    static final String FILE_EXTENSION = ".json";
+
+    private static final String DEFAULT_STATUS_FILE_DIRECTORY = "/home/LogFiles/ApplicationInsights/status";
+
+    @VisibleForTesting
+    static String directory;
+
+    @GuardedBy("StatusFile.class")
```

Comment:
I think better to `@Guard` and synchronize below on an internal field that isn't visible externally, so you don't have to worry about anyone else using the same lock. (even if you have to create the field, e.g. `private final Object lock = new Object()`